libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/barrier.h

```diff
@@ -641,176 +684,441 @@ _LIBCUDACXX_INLINE_VISIBILITY constexpr bool __unused(_Ty...) {return true;}
+        const bool __can_use_complete_tx = __allowed_completions & uint32_t(__completion_mechanism::__mbarrier_complete_tx);
+        _LIBCUDACXX_DEBUG_ASSERT(__can_use_complete_tx == (nullptr != __bar_handle), "Pass non-null bar_handle if and only if can_use_complete_tx.");
+        if _LIBCUDACXX_CONSTEXPR_AFTER_CXX14 (_Align >= 16) {
+            if (__can_use_complete_tx && __isShared(__bar_handle)) {
+                __cp_async_bulk_shared_global(__group, __dest_char, __src_char, __size, __bar_handle);
+                return __completion_mechanism::__mbarrier_complete_tx;
             }
         }
-        else
-        {
-            __strided_memcpy<_Native_alignment>(__destination, __source, __size, __group.thread_rank(), __group.size());
+        // Fallthrough to SM 80..
+    ));
+
+    NV_IF_TARGET(NV_PROVIDES_SM_80, (
+        const bool __can_use_async_group = __allowed_completions & uint32_t(__completion_mechanism::__async_group);
+        if _LIBCUDACXX_CONSTEXPR_AFTER_CXX14 (_Align >= 4) {
+            if (__can_use_async_group) {
+                __cp_async_shared_global_mechanism<_Align>(__group, __dest_char, __src_char, __size);
+                return __completion_mechanism::__async_group;
+            }
         }
+        // Fallthrough..
+    ));
 
-        __memcpy_arrive_on_impl::__arrive_on(__sync, __is_async);
-        , NV_ANY_TARGET,
-            __strided_memcpy<_Native_alignment>(__destination, __source, __size, __group.thread_rank(), __group.size());
-    )
+    __cp_async_fallback_mechanism<_Align>(__group, __dest_char, __src_char, __size);
+    return __completion_mechanism::__sync;
+}
 
-    return __is_async;
+template<_CUDA_VSTD::size_t _Align, typename _Group>
+_LIBCUDACXX_NODISCARD_ATTRIBUTE _LIBCUDACXX_DEVICE inline
+__completion_mechanism __dispatch_memcpy_async_device(_Group const & __group, char * __dest_char, char const * __src_char, _CUDA_VSTD::size_t __size, uint32_t __allowed_completions, uint64_t* __bar_handle) {
+    // Dispatch based on direction of the copy: global to shared, shared to
+    // global, etc.
+
+    // CUDA compilers <= 12.2 may not propagate assumptions about the state space
+    // of pointers correctly. Therefore, we
+    // 1) put the code for each copy direction in a separate function, and
+    // 2) make sure none of the code paths can reach each other by "falling through".
+    //
+    // See nvbug 4074679 and also PR #478.
+    if (__isGlobal(__src_char) && __isShared(__dest_char)) {
+        return __dispatch_memcpy_async_global_to_shared<_Align>(__group, __dest_char, __src_char, __size, __allowed_completions, __bar_handle);
+    } else {
+        return __dispatch_memcpy_async_any_to_any<_Align>(__group, __dest_char, __src_char, __size, __allowed_completions, __bar_handle);
+    }
 }
 
+// __dispatch_memcpy_async is the internal entry point for dispatching to the correct memcpy_async implementation.
+template<typename _Group, typename _Tp, typename _Size>
```

Comment:
This function should not see `_Tp`, it should just be given its alignment; that's the only useful bit of information for this function, so like the original implementation, we should not generate more instantiations of this template than necessary.

Reply:
I'm not convinced avoiding that small duplication (which wouldn't go into all of the entry points, by the way - just the typed ones) is worth overinstantiating this function template.

Reply:
I am with @griwes here, that we should move the `static_assert` to the callers of `__memcpy_async_barrier` and pass void* as well as an additional `alignment` argument to `__memcpy_async_barrier` and `__dispatch_memcpy_async`

Reply:
I have removed the type parameter in commit https://github.com/NVIDIA/cccl/pull/501/commits/38bf8a78442d91232d32747550e0426db8d09c74